### PR TITLE
Add maps concept

### PR DESCRIPTION
**Pull request by smcg468**
# pull request

This issue addresses #2583 

- Renamed the exercise from 'high-score' to 'arcade-high score' to avoid clashing of exercise names
- Added 'maps' to the practises and prerequisites for the 7 exercises listed in the issue
- Changed the prerequisites for the arcade-high-score concept exercise from what was on the elixir track which was

          "lists",
          "tuples",
          "anonymous-functions",
          "default-arguments"

to

           "numbers"
           "strings"

- Could also add an howManyPlayers task to utilise the .size() method if it would help, or maybe best to keep it consistent with the elixir track. 

Let me know of any changes you'd like me to make.
---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/main/POLICIES.md#event-checklist)

### CODE REVIEW

**Comment by sanderploegsma** [comment]
Awesome! I'll try to review this sometime this week 👍

File: exercises/settings.gradle [hunk 1]
```diff
@@ -20,6 +20,7 @@ include 'concept:salary-calculator'
 include 'concept:squeaky-clean'
 include 'concept:tim-from-marketing'
 include 'concept:wizards-and-warriors'
+include 'concept:arcade-high-score'
```
Comment (sanderploegsma) [inline exercises/settings.gradle:23]: The list of exercises is now sorted alphabetically to limit the number of merge conflicts occurring when multiple people are working on new concept exercises at the same time.

Could you move this entry to the appropriate place?

File: concepts/maps/introduction.md
Comment (sanderploegsma) [inline concepts/maps/introduction.md:1]: In my opinion this introduction into Maps should focus more on the basic operations when working with maps. Specifically, I think it should show:

- How to create a new Map instance.
- How to set values using the `put` method.
- How to retrieve values using the `get` method.

To me, the `keySet` and `values` methods are OK to mention in the `about.md`, but should not be introduced until a student knows how to interact with the map contents directly.

File: exercises/concept/arcade-high-score/src/main/java/ArcadeHighScore.java
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/src/main/java/ArcadeHighScore.java:1]: Personally I'd like to change this exercise to be a little bit more object-oriented. I have the following reference implementation in mind, what do you think?

```java
class ArcadeHighScore {

    private Map<String, Integer> highScores = new HashMap<>();

    Map<String, Integer> getHighScores() {
        return highScores;
    }

    void addPlayer(String name, Integer score) {
        highScores.put(name, score);
    }

    void removePlayer(String name) {
        highScores.remove(name);
    }

    void resetScore(String name) {
        highScores.put(name, 0);
    }

    void updateScore(String name, Integer score) {
        Integer oldScore = highScores.getOrDefault(name, 0);
        highScores.put(name, oldScore + score);
    }

    Set<String> getPlayers() {
        return highScores.keySet();
    }
}
```

The first task would then cover defining the private `highScores` field and returning it in the `getHighScores()` method. The other tasks should speak for themselves.

File: exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+public class ArcadeHighScoreTest {
+
+    private final ArcadeHighScore arcadeHighScore = new ArcadeHighScore();
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java:10]: You probably want to create a new instance for every test using `@BeforeEach` to make sure the SUT doesn't retain any state from a previous test, this makes the tests easier to reason about, also for students!

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+In this exercise, you're implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
+
+## 1. Define a new high score map
+
+To make a new high score map, create an empty `HashMap` object by initialising the `Map<String, Integer> highScores` declaration provided.
+
+```java
+System.out.println(ArcadeHighScore.highScores);
+// => {}
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:11]:
```suggestion
ArcadeHighScore arcadeHighScore = new ArcadeHighScore();
arcadeHighScore.getHighScores();
// => {} 
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+To add a player to the high score map, define `ArcadeHighScore.addPlayer()`, which is a function which takes two arguments:
+
+- The first argument is the players name as a String.
+- The second argument is the players score as an Integer.
+
+````java
+ArcadeHighScore.addPlayer("Dave Thomas", 0);
+
+// => {Dave Thomas=0}
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:24]:
```suggestion
arcadeHighScore.addPlayer("Dave Thomas", 0);
arcadeHighScore.getHighScores();
// => {Dave Thomas=0}
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+System.out.println(ArcadeHighScore.highScores);
+// => {} 
+```
+
+## 2. Add players to the high score map
+
+To add a player to the high score map, define `ArcadeHighScore.addPlayer()`, which is a function which takes two arguments:
+
+- The first argument is the players name as a String.
+- The second argument is the players score as an Integer.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:19]:
```suggestion
To add a player to the high score map, implement the `ArcadeHighScore.addPlayer()` method, which takes the player's name and score as arguments.

When called, it should store the player and their score in the `highScores` map.
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+ArcadeHighScore.addPlayer("Dave Thomas", 0);
+
+// => {Dave Thomas=0}
+````
+
+## 3. Remove players from the score map
+
+To remove a player from the high score map, define the `ArcadeHighScore.removePlayer()` method, which takes one argument:
+
+- The argument is the key of the item, in this case the name of the player.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:31]:
```suggestion
To remove a player from the high score map, implement the `ArcadeHighScore.removePlayer()` method, which takes the player's name as an argument.

When called, it should remove the entry associated to the player from the `highScores` map.
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ArcadeHighScore.removePlayer("Dave Thomas");
+
+// => {}
+````
+
+## 4. Reset a player's score
+
+To reset a player's score, define the `ArcadeHighScore.resetScore()` method, which takes 1 argument:
+
+- The argument is the players name.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:43]:
```suggestion
To reset a player's score, implement the `ArcadeHighScore.resetScore()` method, which takes the player's name as an argument.

When called, it should reset the score of the player to `0` in the `highScores` map.
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+## 3. Remove players from the score map
+
+To remove a player from the high score map, define the `ArcadeHighScore.removePlayer()` method, which takes one argument:
+
+- The argument is the key of the item, in this case the name of the player.
+
+````java
+ArcadeHighScore.removePlayer("Dave Thomas");
+
+// => {}
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:36]:
```suggestion
arcadeHighScore.removePlayer("Dave Thomas");
arcadeHighScore.getHighScores();
// => {}
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+## 4. Reset a player's score
+
+To reset a player's score, define the `ArcadeHighScore.resetScore()` method, which takes 1 argument:
+
+- The argument is the players name.
+
+```java
+ArcadeHighScore.resetScore("Dave Thomas");
+// => (Dave Thomas=0);
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:47]:
```suggestion
arcadeHighScore.resetScore("Dave Thomas");
arcadeHighScore.getHighScores();
// => (Dave Thomas=0);
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+## 5. Update a player's score
+
+To update a player's score by adding to the previous score, define `ArcadeHighScore.updateScore()`, which takes two arguments:
+
+- The first argument is the name of the player whose score you wish to update.
+- The second argument is the score that you wish to **add** to their current high score.
+
+The function should also work if the player doesn't have a previous score - assume the previous score is 0.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:57]:
```suggestion
To update a player's score by adding to the previous score, implement `ArcadeHighScore.updateScore()` method, which takes the player's name and their new score as arguments.

When called, it should _add_ the new score to the player's previous score stored in the `highScores` map.
If the player doesn't have a previous score, assume the previous score is `0`.
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+// Adding a players score to update.
+ArcadeHighScore.addPlayer("Lional Messi", 48);
+// => {Lionel Messi=48}
+
+ArcadeHighScore.updateScore("Lionel Messi", 40);
+// => {Lionel Messi=88}
+        
+// Updating a players score who doesn't have a previous score
+ArcadeHighScore.updateScore("Dave Thomas", 57);
+// => {Lionel Messi=88, Dave Thomas=57}
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:69]:
```suggestion
ArcadeHighScore arcadeHighScore = new ArcadeHighScore();

// Adding a players score to update
arcadeHighScore.addPlayer("Lionel Messi", 48);
arcadeHighScore.updateScore("Lionel Messi", 40);
arcadeHighScore.getHighScores();
// => {Lionel Messi=88}

// Updating a players score who doesn't have a previous score
arcadeHighScore.updateScore("Dave Thomas", 57);
arcadeHighScore.getHighScores();
// => {Lionel Messi=88, Dave Thomas=57}
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+## 6. Get a list of players
+
+To get a list of players, define the `ArcadeHighScore.listOfPlayers()`, which takes no arguments:
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:74]:
```suggestion
To get a list of players, implement the `ArcadeHighScore.getPlayers()` method.

When called, it should return a collection containing all player's names currently stored in the `highScores` map.
```

File: exercises/concept/arcade-high-score/.docs/instructions.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+````java
+ArcadeHighScore.listOfPlayers();
+// => {Lionel Messi, Dave Thomas}
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.docs/instructions.md:78]:
```suggestion
arcadeHighScore.getPlayers();
// => {Lionel Messi, Dave Thomas}
```

File: exercises/concept/arcade-high-score/.meta/design.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+- Know how to add key/value pairs in a map.
+- Know how to get values from a map.
+- Know how to get keys from a map.
+- Know how to update values in a map.
+- Know how to remove items from a map.
+- Know how to retrieve all items in a map.
+- Know what a `HashMap` is.
+- Know what a `LinkedHashMap` is.
+- Know what a `TreeMap` is.
+- Know the main differences in `HashMap`, `LinkedHashMap` and `TreeMap`.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.meta/design.md:16]: Note on these objectives: they should cover the learning objectives from the _exercise_, not the _concept_. I believe this exercise doesn't cover the different implementations of the `Map` interface, nor how to retrieve all items in a map.

File: exercises/concept/arcade-high-score/.meta/design.md [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+## Out of scope
+
+- `AbstractMap`
+- `IdentityHashMap`
+- `WeakHashMap`
+- `EnumMap`
+- `ConcurrentHashMap`
+- `ConcurrentSkipListMap`
+- `SortedMap`
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.meta/design.md:26]: We can probably cover all future additions of `Map` implementations in Java by summarizing this as "Other implementations of the `Map` interface"

File: exercises/concept/arcade-high-score/.meta/design.md [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+## Concepts
+
+- `maps`
+
+## Prerequisites
+
+- `strings`: know how to use string formatting.
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/.meta/design.md:34]: Technically I don't think string formatting is used in this exercise, so maybe `strings` doesn't have to be a prerequisite.

File: exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+
+    @Test
+    @Tag("task:1")
+    @DisplayName("Define a map to store the high scores")
+    public void defineMap() {
+        assertNotNull(arcadeHighScore.defineMap());
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java:17]: We should probably (also) check that it's empty.

File: exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        assertThat(arcadeHighScore.highScores).containsEntry(name, score);
+
+        arcadeHighScore.resetScore("David James");
+        assertThat(arcadeHighScore.highScores).containsEntry(name, 0);
+    }
+
+    @Test
+    @Tag("task:5")
+    @DisplayName("Update a players score")
+    public void updateScore() {
```
Comment (sanderploegsma) [inline exercises/concept/arcade-high-score/src/test/java/ArcadeHighScoreTest.java:62]: I think we need to add a second test for this task that verifies the requirement when `updateScore` is called with a player that doesn't exist in the map yet.

**Review by sanderploegsma — CHANGES_REQUESTED** [review]
Thanks for picking this up! It already looks very good, and I left a bunch of comments mainly related to making the exercise a little bit more object-oriented, which is more in line with the Java track's policies.

Could you have a look at them and request a new review when you have addressed them?

**Comment by sanderploegsma** [comment]
@smcg468 Are you still working on this? Leave a comment if you are, otherwise I'm going to close this due to inactivity.

**Comment by smcg468** [comment]
> @smcg468 Are you still working on this? Leave a comment if you are, otherwise I'm going to close this due to inactivity.

Yes still working on this, been quite busy over Christmas but should get back to it soon

**Comment by smcg468** [comment]
@sanderploegsma Sorry yes will get back to it this weekend

**Comment by sanderploegsma** [comment]
Take your time! I didn't mean to rush you, but I figured you may have lost track of this PR 😉

**Comment by manumafe98** [comment]
Hey @smcg468! we will like to know if you need help finishing this PR! it's an interesting concept to be added to the java track, so if you are busy with work we can give you a hand with it! Thanks for your time 😃

**Comment by sanderploegsma** [comment]
@smcg468 I see you've been busy, nice!

Since some of the CI jobs are still failing I'm going to assume that you're still working on addressing the feedback. Just make sure to hit the 're-request review' button on this PR once you're done, so I know it's ready for another look. 😉

**Comment by manumafe98** [comment]
hey @smcg468 do you need some help with this pr?

**Pull request closed without merging**